File: practice_java_program/notpad/LongestSubstringwithoutRepeatingCharacterTwoPointerHashSet.java
import java.util.*;

public class LongestSubstringwithoutRepeatingCharacterTwoPointerHashSet {
    public static void main(String[] args) {

        String s = "abcabcbb";

        int left = 0;
        int right = 0;
        int maxLength = 0;

        HashSet<Character> set = new HashSet<>();

        while (right < s.length()) {		

            char ch = s.charAt(right);		//c

            if (!set.contains(ch)) {
                set.add(ch);			//abc
                maxLength = Math.max(maxLength, right - left + 1);  	//1
                right++;
            } else {
                set.remove(s.charAt(left));
                left++;
            }
        }

        System.out.println("Longest Length = " + maxLength);
    }
}



import java.util.*;

public class Main {
    public static void main(String[] args) {

        String s = "abcabcbb";

        Set<Character> set = new HashSet<>();
        int left = 0;
        int maxLen = 0;

        for (int right = 0; right < s.length(); right++) {  

            while (set.contains(s.charAt(right))) {   
                set.remove(s.charAt(left));   
                left++;
            }

            set.add(s.charAt(right));           
            maxLen = Math.max(maxLen, right - left + 1);
        }

        System.out.println("Longest substring length: " + maxLen);
    }
}

File: python_oriented_interview/notpad/LongestSubstringwithoutRepeatingCharacterTwoPointerHashSet.py
# First method: Two pointer with set logic
def method_1(s):
    left = 0
    right = 0
    maxLength = 0
    char_set = set()

    while right < len(s):
        ch = s[right]
        if ch not in char_set:
            char_set.add(ch)
            maxLength = max(maxLength, right - left + 1)
            right += 1
        else:
            char_set.remove(s[left])
            left += 1

    print("Longest Length =", maxLength)

# Second method: Sliding window logic
def method_2(s):
    char_set = set()
    left = 0
    maxLen = 0

    for right in range(len(s)):
        while s[right] in char_set:
            char_set.remove(s[left])
            left += 1
        char_set.add(s[right])
        maxLen = max(maxLen, right - left + 1)

    print("Longest substring length:", maxLen)

if __name__ == "__main__":
    s = "abcabcbb"
    method_1(s)
    method_2(s)
